File: runtimes/core/src/objects/gcs/bucket.rs
use async_stream::try_stream;
use futures::TryStreamExt;
use google_cloud_storage::http::objects::download::Range;
use google_cloud_storage::http::objects::get::GetObjectRequest;
use google_cloud_storage::http::objects::upload::{Media, UploadObjectRequest, UploadType};
use google_cloud_storage::sign::SignBy;
use google_cloud_storage::sign::SignedURLOptions;
use std::borrow::Cow;
use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;
use std::time::{Duration, SystemTime};
use tokio::io::AsyncRead;
use url::Url;

use crate::encore::runtime::v1 as pb;
use crate::objects::{
    AttrsOptions, DeleteOptions, DownloadOptions, DownloadStream, Error, ExistsOptions, ListEntry,
    ListOptions, ObjectAttrs, PublicUrlError, UploadOptions, UploadUrlOptions,
};
use crate::{objects, CloudName, EncoreName};
use google_cloud_storage as gcs;

use super::LazyGCSClient;

#[derive(Debug)]
pub struct Bucket {
    client: Arc<LazyGCSClient>,
    encore_name: EncoreName,
    cloud_name: CloudName,
    public_base_url: Option<String>,
    key_prefix: Option<String>,
    local_sign: Option<LocalSignOptions>,
}

#[derive(Debug)]
pub struct LocalSignOptions {
    base_url: String,
    access_id: String,
    private_key: String,
}

fn local_sign_config_from_client(client: Arc<LazyGCSClient>) -> Option<LocalSignOptions> {
    client.cfg.local_sign.as_ref().map(|cfg| LocalSignOptions {
        base_url: cfg.base_url.clone(),
        access_id: cfg.access_id.clone(),
        private_key: cfg.private_key.clone(),
    })
}

impl Bucket {
    pub(super) fn new(client: Arc<LazyGCSClient>, cfg: &pb::Bucket) -> Self {
        Self {
            client: client.clone(),
            encore_name: cfg.encore_name.clone().into(),
            cloud_name: cfg.cloud_name.clone().into(),
            public_base_url: cfg.public_base_url.clone(),
            key_prefix: cfg.key_prefix.clone(),
            local_sign: local_sign_config_from_client(client.clone()),
        }
    }

    /// Computes the object name, including the key prefix if present.
    fn obj_name<'a>(&'_ self, name: Cow<'a, str>) -> Cow<'a, str> {
        match &self.key_prefix {
            Some(prefix) => {
                let mut key = prefix.to_owned();
                key.push_str(&name);
                Cow::Owned(key)
            }
            None => name,
        }
    }

    /// Returns the name with the key prefix stripped, if present.
    fn strip_prefix<'a>(&'_ self, name: Cow<'a, str>) -> Cow<'a, str> {
        match &self.key_prefix {
            Some(prefix) => name
                .as_ref()
                .strip_prefix(prefix)
                .map(|s| Cow::Owned(s.to_string()))
                .unwrap_or(name),
            None => name,
        }
    }
}

impl objects::BucketImpl for Bucket {
    fn name(&self) -> &EncoreName {
        &self.encore_name
    }

    fn object(self: Arc<Self>, name: String) -> Arc<dyn objects::ObjectImpl> {
        Arc::new(Object {
            bkt: self,
            key: name,
        })
    }

    fn list(
        self: Arc<Self>,
        options: ListOptions,
    ) -> Pin<Box<dyn Future<Output = Result<objects::ListStream, objects::Error>> + Send + 'static>>
    {
        Box::pin(async move {
            match self.client.get().await {
                Ok(client) => {
                    let client = client.clone();

                    let mut total_seen = 0;
                    const DEFAULT_MAX_RESULTS: u64 = 1000;
                    let s: objects::ListStream = Box::new(try_stream! {
                        let max_results = if let Some(limit) = options.limit {
                            limit.min(DEFAULT_MAX_RESULTS) as i32
                        } else {
                            DEFAULT_MAX_RESULTS as i32
                        };

                        let mut req = gcs::http::objects::list::ListObjectsRequest {
                            bucket: self.cloud_name.to_string(),
                            max_results: Some(max_results),
                            ..Default::default()
                        };


                        // Filter by key prefix, if provided.
                        if let Some(key_prefix) = &self.key_prefix {
                            req.prefix = Some(key_prefix.clone());
                        }

                        'PageLoop:
                        loop {
                            let resp = client.list_objects(&req).await.map_err(|e| Error::Other(e.into()))?;
                            if let Some(items) = resp.items {
                                for obj in items {
                                    total_seen += 1;
                                    if let Some(limit) = options.limit {
                                        if total_seen > limit {
                                            break 'PageLoop;
                                        }
                                    }

                                    let entry = ListEntry {
                                        name: self.strip_prefix(Cow::Owned(obj.name)).into_owned(),
                                        size: obj.size as u64,
                                        etag: obj.etag,
                                    };
                                    yield entry;
                                }
                            }

                            req.page_token = resp.next_page_token;

                            // Are we close to being done? If so, adjust the max_results
                            // to avoid over-fetching.
                            if let Some(limit) = options.limit {
                                if total_seen >= limit {
                                    break 'PageLoop;
                                }
                                let remaining = limit - total_seen;
                                req.max_results = Some(remaining.min(DEFAULT_MAX_RESULTS) as i32);
                            }

                            if req.page_token.is_none() {
                                break;
                            }
                        }
                    });

                    Ok(s)
                }

                Err(err) => Err(Error::Internal(anyhow::anyhow!(
                    "unable to resolve client: {}",
                    err
                ))),
            }
        })
    }
}

#[derive(Debug)]
struct Object {
    bkt: Arc<Bucket>,
    key: String,
}

impl objects::ObjectImpl for Object {
    fn bucket_name(&self) -> &EncoreName {
        &self.bkt.encore_name
    }

    fn key(&self) -> &str {
        &self.key
    }

    fn attrs(
        self: Arc<Self>,
        options: AttrsOptions,
    ) -> Pin<Box<dyn Future<Output = Result<ObjectAttrs, Error>> + Send>> {
        Box::pin(async move {
            match self.bkt.client.get().await {
                Ok(client) => {
                    let mut req = gcs::http::objects::get::GetObjectRequest {
                        bucket: self.bkt.cloud_name.to_string(),
                        object: self.bkt.obj_name(Cow::Borrowed(&self.key)).into_owned(),
                        ..Default::default()
                    };

                    if let Some(version) = options.version {
                        req.generation = Some(parse_version(version)?);
                    }

                    let obj = client.get_object(&req).await.map_err(map_err)?;
                    Ok(ObjectAttrs {
                        name: obj.name,
                        version: Some(obj.generation.to_string()),
                        size: obj.size as u64,
                        content_type: obj.content_type,
                        etag: obj.etag,
                    })
                }
                Err(err) => Err(Error::Internal(anyhow::anyhow!(
                    "unable to resolve client: {}",
                    err
                ))),
            }
        })
    }

    fn signed_upload_url(
        self: Arc<Self>,
        options: UploadUrlOptions,
    ) -> Pin<Box<dyn Future<Output = Result<String, Error>> + Send>> {
        Box::pin(async move {
            match self.bkt.client.get().await {
                Ok(client) => {
                    let gcs_opts = SignedURLOptions {
                        method: gcs::sign::SignedURLMethod::PUT,
                        expires: Duration::from_secs(options.ttl),
                        start_time: Some(SystemTime::now()),
                        ..Default::default()
                    };

                    // We use a fake GCS service for local development. Ideally, the runtime
                    // code would be oblivious to this once the GCS client is set up. But that
                    // turns out to be difficult for URL signing, so we add a special case
                    // here.
                    let local_sign = &self.bkt.local_sign;
                    let access_id: Option<String>;
                    let sign_by: Option<SignBy>;
                    if let Some(opt) = local_sign {
                        access_id = Some(opt.access_id.clone());
                        sign_by = Some(SignBy::PrivateKey(opt.private_key.as_bytes().to_vec()));
                    } else {
                        access_id = None;
                        sign_by = None;
                    }

                    let name = self.bkt.obj_name(Cow::Borrowed(&self.key)).into_owned();
                    let mut url = client
                        .signed_url(&self.bkt.cloud_name, &name, access_id, sign_by, gcs_opts)
                        .await
                        .map_err(map_sign_err)?;

                    // More special handling for the local dev case.
                    if let Some(cfg) = local_sign {
                        url = replace_url_prefix(url, cfg.base_url.clone());
                    }

                    Ok(url)
                }
                Err(err) => Err(Error::Internal(anyhow::anyhow!(
                    "unable to resolve client: {}",
                    err
                ))),
            }
        })
    }

    fn exists(
        self: Arc<Self>,
        options: ExistsOptions,
    ) -> Pin<Box<dyn Future<Output = Result<bool, Error>> + Send>> {
        Box::pin(async move {
            match self.bkt.client.get().await {
                Ok(client) => {
                    let mut req = gcs::http::objects::get::GetObjectRequest {
                        bucket: self.bkt.cloud_name.to_string(),
                        object: self.bkt.obj_name(Cow::Borrowed(&self.key)).into_owned(),
                        ..Default::default()
                    };

                    if let Some(version) = options.version {
                        req.generation = Some(parse_version(version)?);
                    }

                    match client.get_object(&req).await.map_err(map_err) {
                        Ok(_obj) => Ok(true),
                        Err(Error::NotFound) => Ok(false),
                        Err(err) => Err(err),
                    }
                }
                Err(err) => Err(Error::Internal(anyhow::anyhow!(
                    "unable to resolve client: {}",
                    err
                ))),
            }
        })
    }

    fn upload(
        self: Arc<Self>,
        data: Box<dyn AsyncRead + Unpin + Send + Sync + 'static>,
        opts: objects::UploadOptions,
    ) -> Pin<Box<dyn Future<Output = Result<ObjectAttrs, Error>> + Send>> {
        Box::pin(async move {
            match self.bkt.client.get().await {
                Ok(client) => {
                    let mut req = UploadObjectRequest {
                        bucket: self.bkt.cloud_name.to_string(),
                        ..Default::default()
                    };

                    let cloud_name = self.bkt.obj_name(Cow::Borrowed(&self.key));
                    let mut media = Media::new(cloud_name.into_owned());

                    apply_upload_opts(opts, &mut req, &mut media);

                    let upload_type = UploadType::Simple(media);
                    let stream = tokio_util::io::ReaderStream::new(data);

                    match client
                        .upload_streamed_object(&req, stream, &upload_type)
                        .await
                    {
                        Ok(obj) => Ok(ObjectAttrs {
                            name: obj.name,
                            version: Some(obj.generation.to_string()),
                            size: obj.size as u64,
                            content_type: obj.content_type,
                            etag: obj.etag,
                        }),
                        Err(err) => Err(map_err(err)),
                    }
                }
                Err(err) => Err(Error::Internal(anyhow::anyhow!(
                    "unable to resolve client: {}",
                    err
                ))),
            }
        })
    }

    fn download(
        self: Arc<Self>,
        options: DownloadOptions,
    ) -> Pin<Box<dyn Future<Output = Result<DownloadStream, Error>> + Send>> {
        fn convert_err(err: gcs::http::Error) -> Error {
            use gcs::http::error::ErrorResponse;
            match err {
                gcs::http::Error::Response(ErrorResponse { code: 404, .. }) => Error::NotFound,
                gcs::http::Error::HttpClient(err)
                    if err.status().map(|s| s.as_u16()) == Some(404) =>
                {
                    Error::NotFound
                }
                err => Error::Other(err.into()),
            }
        }

        Box::pin(async move {
            match self.bkt.client.get().await {
                Ok(client) => {
                    let mut req = GetObjectRequest {
                        bucket: self.bkt.cloud_name.to_string(),
                        object: self.bkt.obj_name(Cow::Borrowed(&self.key)).into_owned(),
                        ..Default::default()
                    };

                    if let Some(version) = options.version {
                        req.generation = Some(parse_version(version)?);
                    }

                    let resp = client
                        .download_streamed_object(&req, &Range::default())
                        .await;

                    let stream = resp.map_err(convert_err)?;
                    let stream: DownloadStream = Box::pin(stream.map_err(convert_err));
                    Ok(stream)
                }
                Err(err) => Err(Error::Internal(anyhow::anyhow!(
                    "unable to resolve client: {}",
                    err
                ))),
            }
        })
    }

    fn delete(
        self: Arc<Self>,
        options: DeleteOptions,
    ) -> Pin<Box<dyn Future<Output = Result<(), Error>> + Send>> {
        Box::pin(async move {
            match self.bkt.client.get().await {
                Ok(client) => {
                    let mut req = gcs::http::objects::delete::DeleteObjectRequest {
                        bucket: self.bkt.cloud_name.to_string(),
                        object: self.bkt.obj_name(Cow::Borrowed(&self.key)).into_owned(),
                        ..Default::default()
                    };

                    if let Some(version) = options.version {
                        req.generation = Some(parse_version(version)?);
                    }

                    match client.delete_object(&req).await.map_err(map_err) {
                        Ok(_) => Ok(()),
                        Err(Error::NotFound) => Err(Error::NotFound),
                        Err(err) => Err(err),
                    }
                }
                Err(err) => Err(Error::Internal(anyhow::anyhow!(
                    "unable to resolve client: {}",
                    err
                ))),
            }
        })
    }

    fn public_url(&self) -> Result<String, PublicUrlError> {
        let Some(base_url) = self.bkt.public_base_url.clone() else {
            return Err(PublicUrlError::PrivateBucket);
        };

        let url = objects::public_url(base_url, &self.key);
        Ok(url)
    }
}

fn replace_url_prefix(orig_url: String, base: String) -> String {
    match Url::parse(&orig_url) {
        Ok(url) => {
            let mut out = format!(
                "{}/{}",
                base.trim_end_matches('/'),
                url.path().trim_start_matches("/")
            );
            if let Some(query) = url.query() {
                out.push('?');
                out.push_str(query);
            }
            out
        }
        Err(_) => {
            // If the input URL fails parsing, just don't do the replace
            orig_url
        }
    }
}

fn apply_upload_opts(opts: UploadOptions, req: &mut UploadObjectRequest, media: &mut Media) {
    if let Some(content_type) = opts.content_type {
        media.content_type = Cow::Owned(content_type);
    }
    if let Some(pre) = opts.preconditions {
        if pre.not_exists == Some(true) {
            req.if_generation_match = Some(0);
        }
    }
}

fn parse_version(version: String) -> Result<i64, Error> {
    version.parse().map_err(|err| {
        Error::Other(anyhow::anyhow!(
            "invalid version number {}: {}",
            version,
            err
        ))
    })
}

fn map_err(err: gcs::http::Error) -> Error {
    use gcs::http::error::ErrorResponse;
    match err {
        gcs::http::Error::Response(ErrorResponse { code, .. }) => match code {
            404 => Error::NotFound,
            412 => Error::PreconditionFailed,
            _ => Error::Other(err.into()),
        },
        gcs::http::Error::HttpClient(err) => {
            let status = err.status().map(|s| s.as_u16());
            match status {
                Some(404) => Error::NotFound,
                Some(412) => Error::PreconditionFailed,
                _ => Error::Other(err.into()),
            }
        }
        err => Error::Other(err.into()),
    }
}

fn map_sign_err(err: gcs::sign::SignedURLError) -> Error {
    match err {
        gcs::sign::SignedURLError::InvalidOption(_e) => Error::PreconditionFailed,
        err => Error::Internal(anyhow::anyhow!(err)),
    }
}

File: runtimes/go/storage/objects/internal/providers/gcs/bucket.go
package gcs

import (
	"context"
	"errors"
	"fmt"
	"iter"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"cloud.google.com/go/storage"
	"google.golang.org/api/googleapi"
	"google.golang.org/api/iterator"
	"google.golang.org/api/option"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"

	"encore.dev/appruntime/exported/config"
	"encore.dev/storage/objects/internal/types"
)

type Manager struct {
	ctx     context.Context
	runtime *config.Runtime
	clients map[*config.BucketProvider]*storage.Client
}

func NewManager(ctx context.Context, runtime *config.Runtime) *Manager {
	return &Manager{ctx: ctx, runtime: runtime, clients: make(map[*config.BucketProvider]*storage.Client)}
}

type localSignOptions struct {
	baseURL    string
	accessID   string
	privateKey string
}

type bucket struct {
	client    *storage.Client
	cfg       *config.Bucket
	handle    *storage.BucketHandle
	localSign *localSignOptions
}

func (mgr *Manager) ProviderName() string { return "gcs" }

func (mgr *Manager) Matches(cfg *config.BucketProvider) bool {
	return cfg.GCS != nil
}

func (mgr *Manager) NewBucket(provider *config.BucketProvider, runtimeCfg *config.Bucket) types.BucketImpl {
	client := mgr.clientForProvider(provider)

	localSign := localSignOptionsForProvider(provider)
	handle := client.Bucket(runtimeCfg.CloudName)
	return &bucket{client, runtimeCfg, handle, localSign}
}

func (b *bucket) Download(data types.DownloadData) (types.Downloader, error) {
	obj := b.handle.Object(data.Object.String())
	if data.Version != "" {
		if gen, err := strconv.ParseInt(data.Version, 10, 64); err == nil {
			obj = obj.Generation(gen)
		}
	}
	r, err := obj.NewReader(data.Ctx)
	return r, mapErr(err)
}

func (b *bucket) Upload(data types.UploadData) (types.Uploader, error) {
	ctx, cancel := context.WithCancelCause(data.Ctx)
	obj := b.handle.Object(data.Object.String())

	if data.Pre.NotExists {
		obj = obj.If(storage.Conditions{
			DoesNotExist: true,
		})
	}

	w := obj.NewWriter(ctx)
	w.ContentType = data.Attrs.ContentType

	u := &uploader{
		cancel: cancel,
		w:      w,
	}
	return u, nil
}

type uploader struct {
	cancel context.CancelCauseFunc
	w      *storage.Writer
}

func (u *uploader) Write(p []byte) (int, error) {
	n, err := u.w.Write(p)
	return n, mapErr(err)
}

func (u *uploader) Complete() (*types.ObjectAttrs, error) {
	if err := u.w.Close(); err != nil {
		return nil, mapErr(err)
	}

	attrs := u.w.Attrs()
	return mapAttrs(attrs), nil
}

func (u *uploader) Abort(err error) {
	u.cancel(err)
}

func mapAttrs(attrs *storage.ObjectAttrs) *types.ObjectAttrs {
	if attrs == nil {
		return nil
	}
	return &types.ObjectAttrs{
		Object:      types.CloudObject(attrs.Name),
		Version:     strconv.FormatInt(attrs.Generation, 10),
		ContentType: attrs.ContentType,
		Size:        attrs.Size,
		ETag:        attrs.Etag,
	}
}

func mapListEntry(attrs *storage.ObjectAttrs) *types.ListEntry {
	return &types.ListEntry{
		Object: types.CloudObject(attrs.Name),
		Size:   attrs.Size,
		ETag:   attrs.Etag,
	}
}

func (b *bucket) List(data types.ListData) iter.Seq2[*types.ListEntry, error] {
	iter := b.handle.Objects(data.Ctx, &storage.Query{
		Prefix: data.Prefix,
	})
	var n int64
	return func(yield func(*types.ListEntry, error) bool) {
		for {
			res, err := iter.Next()
			if err == iterator.Done {
				return
			}

			// Are we over the limit?
			if data.Limit != nil && n >= *data.Limit {
				return
			}
			n++

			var entry *types.ListEntry
			if res != nil {
				entry = mapListEntry(res)
			}

			if !yield(entry, mapErr(err)) {
				return
			}
		}
	}
}

func (b *bucket) Remove(data types.RemoveData) error {
	obj := b.handle.Object(data.Object.String())

	if data.Version != "" {
		if gen, err := strconv.ParseInt(data.Version, 10, 64); err == nil {
			obj = obj.Generation(gen)
		}
	}

	err := obj.Delete(data.Ctx)
	return mapErr(err)
}

func (b *bucket) Attrs(data types.AttrsData) (*types.ObjectAttrs, error) {
	obj := b.handle.Object(data.Object.String())

	if data.Version != "" {
		if gen, err := strconv.ParseInt(data.Version, 10, 64); err == nil {
			obj = obj.Generation(gen)
		}
	}

	resp, err := obj.Attrs(data.Ctx)
	return mapAttrs(resp), mapErr(err)
}

func (b *bucket) SignedUploadURL(data types.UploadURLData) (string, error) {
	opts := &storage.SignedURLOptions{
		Scheme:  storage.SigningSchemeV4,
		Method:  "PUT",
		Expires: time.Now().Add(data.TTL),
	}
	return b.signedURL(data.Object.String(), opts)
}

func (b *bucket) SignedDownloadURL(data types.DownloadURLData) (string, error) {
	opts := &storage.SignedURLOptions{
		Scheme:  storage.SigningSchemeV4,
		Method:  "GET",
		Expires: time.Now().Add(data.TTL),
	}
	return b.signedURL(data.Object.String(), opts)
}

func (b *bucket) signedURL(object string, opts *storage.SignedURLOptions) (string, error) {
	// We use a fake GCS service for local development. Ideally, the runtime
	// code would be oblivious to this once the GCS client is set up. But that
	// turns out to be difficult for URL signing, so we add a special case
	// here.
	if b.localSign != nil {
		opts.GoogleAccessID = b.localSign.accessID
		opts.PrivateKey = []byte(b.localSign.privateKey)
	}

	url, err := b.handle.SignedURL(object, opts)
	if err != nil {
		return "", mapErr(err)
	}

	// More special handling for the local dev case.
	if b.localSign != nil {
		url = replaceURLPrefix(url, b.localSign.baseURL)
	}

	return url, nil
}

func replaceURLPrefix(origUrl string, base string) string {
	u, err := url.Parse(origUrl)
	if err != nil {
		return origUrl // If the input URL is not valid, just return it as-is
	}
	out := base
	if u.Path != "" {
		out = strings.TrimRight(out, "/") + "/" + strings.TrimLeft(u.Path, "/")
	}
	if u.RawQuery != "" {
		out += "?" + u.RawQuery
	}
	return out
}

func (mgr *Manager) clientForProvider(prov *config.BucketProvider) *storage.Client {
	if client, ok := mgr.clients[prov]; ok {
		return client
	}

	var opts []option.ClientOption
	if prov.GCS.Anonymous {
		opts = append(opts, option.WithoutAuthentication())
	}
	if prov.GCS.Endpoint != "" {
		opts = append(opts, option.WithEndpoint(prov.GCS.Endpoint))
	}

	client, err := storage.NewClient(mgr.ctx, opts...)
	if err != nil {
		panic(fmt.Sprintf("failed to create object storage client: %s", err))
	}

	mgr.clients[prov] = client
	return client
}

func localSignOptionsForProvider(prov *config.BucketProvider) *localSignOptions {
	if opts := prov.GCS.LocalSign; opts == nil {
		return nil
	} else {
		return &localSignOptions{
			baseURL:    opts.BaseURL,
			accessID:   opts.AccessID,
			privateKey: opts.PrivateKey,
		}
	}
}

func mapErr(err error) error {
	switch {
	case err == nil:
		return nil
	case errors.Is(err, storage.ErrObjectNotExist):
		return types.ErrObjectNotExist
	default:
		// Handle precondition failures
		{
			var e *googleapi.Error
			if ok := errors.As(err, &e); ok && e.Code == http.StatusPreconditionFailed {
				return types.ErrPreconditionFailed
			}
		}

		{
			if s, ok := status.FromError(err); ok && s.Code() == codes.AlreadyExists || s.Code() == codes.FailedPrecondition {
				return types.ErrPreconditionFailed
			}
		}

		return err
	}
}
